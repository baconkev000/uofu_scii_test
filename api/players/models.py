from django.db import models

class Player(models.Model):
    """
    Player
    A class model for a player
    """
    Name = models.CharField(max_length=255)
    Nationality = models.CharField(max_length=50)
    National_Position = models.CharField(max_length=5)
    National_Kit = models.PositiveIntegerField()
    Club = models.CharField(max_length=255)
    Club_Position = models.CharField(max_length=5)
    Club_Kit = models.PositiveIntegerField()
    Club_Joining = models.CharField()
    Contract_Expiry = models.PositiveIntegerField()
    Rating = models.PositiveIntegerField()
    Height = models.CharField(max_length=10)
    Weight = models.CharField(max_length=10)
    Preffered_Foot = models.CharField(max_length=10)
    Birth_Date = models.CharField()
    Age = models.PositiveIntegerField()
    Preffered_Position = models.CharField(max_length=10)
    Work_Rate = models.CharField(max_length=20)
    Weak_foot = models.PositiveIntegerField()
    Skill_Moves = models.PositiveIntegerField()
    Ball_Control = models.PositiveIntegerField()
    Dribbling = models.PositiveIntegerField()
    Marking = models.PositiveIntegerField()
    Sliding_Tackle = models.PositiveIntegerField()
    Standing_Tackle = models.PositiveIntegerField()
    Aggression = models.PositiveIntegerField()
    Reactions = models.PositiveIntegerField()
    Attacking_Position = models.PositiveIntegerField()
    Interceptions = models.PositiveIntegerField()
    Vision = models.PositiveIntegerField()
    Composure = models.PositiveIntegerField()
    Crossing = models.PositiveIntegerField()
    Short_Pass = models.PositiveIntegerField()
    Long_Pass = models.PositiveIntegerField()
    Acceleration = models.PositiveIntegerField()
    Speed = models.PositiveIntegerField()
    Stamina = models.PositiveIntegerField()
    Strength = models.PositiveIntegerField()
    Balance = models.PositiveIntegerField()
    Agility = models.PositiveIntegerField()
    Jumping = models.PositiveIntegerField()
    Heading = models.PositiveIntegerField()
    Shot_Power = models.PositiveIntegerField()
    Finishing = models.PositiveIntegerField()
    Long_Shots = models.PositiveIntegerField()
    Curve = models.PositiveIntegerField()
    Freekick_Accuracy = models.PositiveIntegerField()
    Penalties = models.PositiveIntegerField()
    Volleys = models.PositiveIntegerField()
    GK_Positioning = models.PositiveIntegerField()
    GK_Diving = models.PositiveIntegerField()
    GK_Kicking = models.PositiveIntegerField()
    GK_Handling = models.PositiveIntegerField()
    GK_Reflexes = models.PositiveIntegerField()

    class Meta:
        verbose_name_plural = "Players"

    def __str__(self):
        return self.Name
