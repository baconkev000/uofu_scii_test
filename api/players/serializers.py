from rest_framework import serializers
from .models import Player
from datetime import datetime

class PlayerSerializer(serializers.ModelSerializer):

    class Meta:
        model = Player
        fields = '__all__'