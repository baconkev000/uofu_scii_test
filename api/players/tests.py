from .models import Player
from api.users.models import User
from django.core.files.uploadedfile import SimpleUploadedFile
from rest_framework import status
from rest_framework.test import APITestCase
#from rest_framework_simplejwt.tokens import RefreshToken


class PlayerTestCase(APITestCase):
    """Tests related to the Player model."""

    fixtures = ["soccer_small.json"]

    def setUp(self):
        """Set up required items for the tests."""
        super().setUp()

        self.user = User.objects.create_user(
            username="test",
            email="test@test.com",
            password="password",
        )

        self.superuser = User.objects.create_superuser(
            username="test_super",
            email="super@test.com",
            password="password",
        )

    def tearDown(self):
        """Delete things in the reverse of the order they were made"""
        self.superuser.delete()
        self.user.delete()

    def test_create_user(self):
        """Our user in setup is a user"""
        self.assertIsInstance(self.user, User)
        self.assertFalse(self.user.is_superuser)

    def test_get_player_list(self):
        """Can get list of all players"""
        player_list_response = self.client.get(
            "/api/players/",
            format="json",
        )
        self.assertEqual(player_list_response.status_code, status.HTTP_200_OK)
    
    # def test_get_player_by_name(self):
    #     """Can get a player by name"""
    #     player_list_response = self.client.get(
    #         "/api/players/Cristiano%20Ronaldo/",
    #         format="json",
    #     )
    #     self.assertEqual(player_list_response.status_code, status.HTTP_200_OK)

    # def test_get_clubs_list(self):
    #     """Can get list of all players"""
    #     clubs_list_response = self.client.get(
    #         "/api/clubs/",
    #         format="json",
    #     )
    #     self.assertEqual(clubs_list_response.status_code, status.HTTP_200_OK)

    # def test_get_attributes_list(self):
    #     """Can get list of all players"""
    #     attributes_list_response = self.client.get(
    #         "/api/attributes/",
    #         format="json",
    #     )
    #     self.assertEqual(attributes_list_response.status_code, status.HTTP_200_OK)