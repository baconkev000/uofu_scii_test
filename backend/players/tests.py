from api.users.models import User
from rest_framework import status
from rest_framework.test import APITestCase, APIClient
from rest_framework.authtoken.models import Token


class PlayerTestCase(APITestCase):
    """Tests related to the Player model."""

    fixtures = ["soccer_small.json"]

    def setUp(self):
        """Set up required items for the tests."""
        self.client = APIClient()
        self.user = User.objects.create_user(
            username="test",
            email="test@test.com",
            password="password",
        )

        self.superuser = User.objects.create_superuser(
            username="test_super",
            email="super@test.com",
            password="password",
        )
        self.token = Token.objects.create(user=self.user)

    def tearDown(self):
        """Delete things in the reverse of the order they were made"""
        self.superuser.delete()
        self.user.delete()

    def test_create_user(self):
        """Our user in setup is a user"""
        self.assertIsInstance(self.user, User)
        self.assertFalse(self.user.is_superuser)

    def test_get_player_list(self):
        """Can get list of all players"""
        self.client.force_login(user=self.user)
        player_list_response = self.client.get(
            "/api/players/",
            format="json",
        )
        self.assertEqual(player_list_response.status_code, status.HTTP_200_OK)

    def test_cannot_get_player_list_without_token(self):
        """Cannnot get list of all players without an auth token"""
        player_list_response = self.client.get(
            "/api/players/",
            format="json",
        )
        self.assertEqual(player_list_response.status_code, status.HTTP_403_FORBIDDEN)

    def test_get_player_by_name(self):
        """Can get a player by name"""
        self.client.force_login(user=self.user)
        player_list_response = self.client.get(
            "/api/players/Cristiano%20Ronaldo/",
            format="json",
        )
        self.assertEqual(player_list_response.status_code, status.HTTP_200_OK)

    def test_get_clubs_list(self):
        """Can get list of all clubs"""
        self.client.force_login(user=self.user)
        clubs_list_response = self.client.get(
            "/api/clubs/",
            format="json",
        )
        self.assertEqual(clubs_list_response.status_code, status.HTTP_200_OK)

    def test_get_attributes_list(self):
        """Can get list of all attribute names"""
        self.client.force_login(user=self.user)
        attributes_list_response = self.client.get(
            "/api/attributes/",
            format="json",
        )
        self.assertEqual(attributes_list_response.status_code, status.HTTP_200_OK)
