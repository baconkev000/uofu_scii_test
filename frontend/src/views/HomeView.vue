<template>
  <div class="w-full flex space-between bg-slate-300 lg:flex-row flex-col">
    <div
      class="xl:w-1/2 w-full xl:flex-col flex-row px-4 h-max overflow-x-scroll"
    >
      <!-- Dropdown -->
      <div class="flex flex-row justify-end w-full py-8">
        <div class="relative inline-block text-left">
          <div>
            <button
              type="button"
              class="inline-flex w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
              id="menu-button"
              aria-expanded="true"
              aria-haspopup="true"
              @click="toggleDropdown()"
            >
              Attributes
              <svg
                class="-mr-1 h-5 w-5 text-gray-400"
                viewBox="0 0 20 20"
                fill="currentColor"
                aria-hidden="true"
              >
                <path
                  fill-rule="evenodd"
                  d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
                  clip-rule="evenodd"
                />
              </svg>
            </button>
          </div>
          <div
            v-show="isDropdownOpen"
            @click="toggleDropdown"
            class="fixed inset-0 bg-black opacity-25 z-9"
          ></div>
          <div
            v-show="isDropdownOpen"
            class="absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none"
            role="menu"
            aria-orientation="vertical"
            aria-labelledby="menu-button"
            tabindex="-1"
          >
            <div class="py-1" role="none">
              <label
                v-for="attribute in attributes"
                :key="attribute"
                class="flex items-center px-4 py-2 text-sm"
              >
                <input
                  type="checkbox"
                  :value="attribute"
                  v-model="selectedAttributes"
                  class="mr-2"
                  @click="updateAttrs(attribute)"
                />
                {{ attribute }}
              </label>
            </div>
          </div>
        </div>
      </div>
      <!-- Table -->
      <div
        class="w-full mx-auto overflow-x-auto overflow-y-auto h-screen shadow-lg"
      >
        <table class="w-full text-sm text-center rtl:text-right text-black">
          <thead class="text-xs text-gray-700 uppercase bg-gray-600 text-white">
            <tr>
              <th
                @click="updateSorted('Name')"
                scope="col"
                class="px-6 py-3 cursor-pointer"
              >
                Name
              </th>
              <th
                v-for="row in selectedAttributes"
                :key="row"
                @click="updateSorted(row)"
                scope="col"
                class="px-6 py-3 cursor-pointer"
              >
                {{ row }}
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="bg-white border-b bg-slate-200 even:bg-gray-100 odd:bg-gray-200 hover:bg-white cursor-pointer"
              v-for="player in sortedPlayers"
              :key="player.pk"
            >
              <td
                scope="row"
                class="px-6 py-4 font-medium text-gray-900 whitespace-nowrap font-medium"
                @click="getSelectedPlayer(player.Name)"
              >
                {{ player.Name }}
              </td>
              <td
                v-for="attr in selectedAttributes"
                :key="attr"
                class="px-6 py-4"
              >
                {{ player[attr] ? player[attr] : "N/A" }}
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="lg:w-1/2 w-full flex flex-col justify-center items-center bg-slate-200 px-4"
    >
      <PlayerStats
        v-if="showPlayerStats"
        :player="selectedPlayer"
      ></PlayerStats>
      <PlayerVisuals
        v-else
        :players="sortedPlayers"
        :selectedLabel="selectedLabel"
      ></PlayerVisuals>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Options } from "vue-class-component";
import { APIService } from "@/services/api";
import PlayerVisuals from "@/components/PlayerVisual.vue";
import PlayerStats from "@/components/PlayerStats.vue";
import { PlayerFields } from "@/types/player_fields";

@Options({
  components: {
    PlayerVisuals,
    PlayerStats,
  },
})
export default class HomeView extends Vue {
  selectedAttributes: string[] = [
    "Nationality",
    "National_Position",
    "Club",
    "Height",
    "Preffered_Foot",
    "Ball_Control",
    "Dribbling",
    "Vision",
  ];
  sortedPlayers: PlayerFields[] = [];
  attributes: string[] = [];
  isDropdownOpen = false;
  selectedLabel = "Rating";
  showPlayerStats = false;
  selectedPlayer: any = {};

  toggleDropdown() {
    this.isDropdownOpen = !this.isDropdownOpen;
  }

  updateAttrs(attr: string): void {
    const index = this.selectedAttributes.indexOf(attr);

    if (index !== -1) {
      // If the attribute is already selected, remove it
      this.selectedAttributes.splice(index, 1);
    } else {
      // If the attribute is not selected, add it
      this.selectedAttributes.push(attr);
    }
  }

  updateSorted(header: string): void {
    this.selectedLabel = header !== "Name" ? header : "Rating";
    this.showPlayerStats = false;
    header = header.split(" ").join("_");
    this.sortedPlayers = this.sortedPlayers.sort(
      (playerA: PlayerFields, playerB: PlayerFields) => {
        const aValue = playerA[header as keyof typeof playerA];
        const bValue = playerB[header as keyof typeof playerA];

        const compareValues = (aValue: any, bValue: any): number => {
          // If both values are null, consider them equal
          if (aValue === null && bValue === null) {
            return 0;
          }

          // If only aValue is null, prioritize it to the bottom
          if (aValue === null) {
            return 1;
          }

          // If only bValue is null, prioritize it to the bottom
          if (bValue === null) {
            return -1;
          }

          // The rest of your comparison logic for non-null values
          return aValue.localeCompare(bValue);
        };

        const compareNumbers = (aValue: number, bValue: number) => {
          return aValue - bValue;
        };

        if (typeof aValue === "string" && typeof bValue === "string") {
          return aValue.localeCompare(bValue);
        } else if (typeof aValue === "number" && typeof bValue === "number") {
          return compareNumbers(aValue, bValue);
        } else {
          return compareValues(aValue, bValue);
        }
      }
    );
  }
  getSelectedPlayer(playerName: string) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    this.selectedPlayer = this.sortedPlayers.find(
      (p: PlayerFields) => p.Name === playerName
    )!;
    this.showPlayerStats = true;
  }

  async mounted() {
    // get players
    try {
      const response = await APIService.getPlayers();
      this.sortedPlayers = response.data;
    } catch (error) {
      console.log(error);
    }
    // get attributes
    try {
      const response = await APIService.getAttributes();
      this.attributes = response.data;
    } catch (error) {
      console.log(error);
    }
  }
}
</script>
