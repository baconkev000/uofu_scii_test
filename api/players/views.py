from django.shortcuts import render
from rest_framework import permissions, viewsets
from .models import Player

from .serializers import PlayerSerializer
from rest_framework.response import Response
from rest_framework import generics



class PlayerViewSet(viewsets.ReadOnlyModelViewSet):
    """
    A simple viewset for listing players
    """
    queryset = Player.objects.all()
    serializer_class = PlayerSerializer
    
    
class PlayerNameList(generics.ListAPIView):
    queryset = Player.objects.all()
    serializer_class = PlayerSerializer

    def list(self, request, *args, **kwargs):
        """
        A simple list view for listing all player names
        """
        queryset = self.get_queryset()
        serializer = self.get_serializer(queryset, many=True)
        return Response([item['Name'] for item in serializer.data])
    
class PlayerByName(generics.ListAPIView):
    serializer_class = PlayerSerializer
    
    def get_queryset(self):
        """
        A simple view for retrieving players by name
        """
        playerName = self.kwargs['name']
        return Player.objects.filter(Name=playerName)
    
class ClubList(generics.ListAPIView):
    serializer_class = PlayerSerializer
    
    def get_queryset(self):
            return Player.objects.all()

    def list(self, request, *args, **kwargs):
        queryset = self.get_queryset()
        serializer = self.get_serializer(queryset, many=True)

        # Group players by club
        clubs_dict = {}
        for player_data in serializer.data:
            club_name = player_data['Club']
            player_name = player_data['Name']

            if club_name not in clubs_dict:
                clubs_dict[club_name] = {'name': club_name, 'players': []}

            clubs_dict[club_name]['players'].append(player_name)

        clubs_list = [club_info for club_info in clubs_dict.values()]
        return Response(clubs_list)
    
class AttributeNameList(generics.ListAPIView):
    queryset = Player.objects.all()
    serializer_class = PlayerSerializer

    def list(self, request, *args, **kwargs):
        """
        A simple list view for listing all player names
        """
        queryset = self.get_queryset()
        serializer = self.get_serializer(queryset, many=True)
        return Response([item['Attribute'] for item in serializer.data])