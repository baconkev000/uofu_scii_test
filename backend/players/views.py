from django.shortcuts import get_object_or_404
from rest_framework import permissions, viewsets
from rest_framework.authentication import SessionAuthentication, BasicAuthentication
from rest_framework.response import Response
from rest_framework import generics

from .models import Player

from .serializers import PlayerSerializer


class PlayerViewSet(viewsets.ReadOnlyModelViewSet):
    """
    A Simple Viewset for listing players and retreiving a player by name
    """

    queryset = Player.objects.all()
    serializer_class = PlayerSerializer
    # add SessionAuthentication & BasicAuthentication for auth token requirement
    authentication_classes = []
    # add permissions.IsAuthenticate for auth token requirement
    permission_classes = []
    lookup_field = "Name"


class ClubListView(generics.ListAPIView):
    """
    A simple list view for listing clubs and the associated players
    """

    serializer_class = PlayerSerializer

    def get_queryset(self):
        return Player.objects.all()

    def list(self, request):
        queryset = self.get_queryset()
        serializer = self.get_serializer(queryset, many=True)

        # Group players by club
        clubs_dict = {}
        i = 0
        for player_data in serializer.data:
            club_name = player_data["Club"]
            player_name = player_data["Name"]

            if club_name not in clubs_dict:
                i += 1
                clubs_dict[club_name] = {"id": i, "name": club_name, "players": []}

            clubs_dict[club_name]["players"].append(player_name)

        clubs_list = [club_info for club_info in clubs_dict.values()]
        return Response(clubs_list)


class PlayerAttributesListView(generics.ListAPIView):
    """
    A simple list view for listing player attribute names
    """

    queryset = Player.objects.all()
    serializer_class = PlayerSerializer

    def list(self, request):
        queryset = self.get_queryset()
        serializer = self.get_serializer(queryset.first())  # Get the first player's data

        # Exclude specific fields from attribute names
        excluded_fields = ["id", "Name"]
        attribute_names = [key for key in serializer.data.keys() if key not in excluded_fields]

        return Response(attribute_names)
