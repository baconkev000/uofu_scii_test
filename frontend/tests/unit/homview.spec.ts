// Import Vue Test Utils and the component being tested
import { mount } from "@vue/test-utils";
import HomeView from "@/views/HomeView.vue";

import players from "@/fixtures/players.json";
import attrs from "@/fixtures/attributes.json";
import axios from "axios";

// Mock Axios
jest.mock("axios");

describe("HomeView.vue", () => {
  it("renders correctly", async () => {
    // Mock data for testing
    const mockPlayers = players;

    const mockAttributes = attrs;

    // Mock Axios response
    (axios.get as jest.Mock).mockResolvedValueOnce({ data: mockPlayers });
    (axios.get as jest.Mock).mockResolvedValueOnce({ data: mockAttributes });

    // Mount the component
    const wrapper = mount(HomeView);

    // Wait for async data to be fetched
    await wrapper.vm.$nextTick();

    // Check if the component renders correctly
    expect(wrapper.html()).toMatchSnapshot();
  });
});
