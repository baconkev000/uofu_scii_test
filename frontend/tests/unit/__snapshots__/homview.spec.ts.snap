// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeView.vue renders correctly 1`] = `
<div class="w-full flex-col">
  <!-- Dropdown -->
  <div class="flex flex-row justify-end w-5/6 py-8">
    <div class="relative inline-block text-left">
      <div><button type="button" class="inline-flex w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50" id="menu-button" aria-expanded="true" aria-haspopup="true"> Attributes <svg class="-mr-1 h-5 w-5 text-gray-400" viewBox="0 0 20 20" fill="currentColor" aria-hidden="true">
            <path fill-rule="evenodd" d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z" clip-rule="evenodd"></path>
          </svg></button></div>
      <div class="fixed inset-0 bg-black opacity-25 z-9" style="display: none;"></div>
      <div class="absolute right-0 z-10 mt-2 w-56 origin-top-right rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none" role="menu" aria-orientation="vertical" aria-labelledby="menu-button" tabindex="-1" style="display: none;">
        <div class="py-1" role="none"></div>
      </div>
    </div>
  </div><!-- Table -->
  <div class="w-3/4 mx-auto overflow-x-auto shadow-lg">
    <table class="w-full text-sm text-center rtl:text-right text-black">
      <thead class="text-xs text-gray-700 uppercase bg-gray-600 text-white">
        <tr>
          <th scope="col" class="px-6 py-3"> Name </th>
          <th scope="col" class="px-6 py-3">Nationality</th>
          <th scope="col" class="px-6 py-3">National_Position</th>
          <th scope="col" class="px-6 py-3">Club</th>
          <th scope="col" class="px-6 py-3">Height</th>
          <th scope="col" class="px-6 py-3">Preffered_Foot</th>
          <th scope="col" class="px-6 py-3">Ball_Control</th>
          <th scope="col" class="px-6 py-3">Dribbling</th>
          <th scope="col" class="px-6 py-3">Vision</th>
        </tr>
      </thead>
      <tbody></tbody>
    </table>
  </div>
</div>
`;
